src/psyclone/tests/dynamo0p3_basis_test.py

```diff
@@ -1742,3 +1730,56 @@ def test_diff_basis_unsupp_space():
         _ = kernel.gen_stub
     assert 'Unsupported space for differential basis function' \
         in str(excinfo.value)
+
+
+def test_dynbasisfns_unsupp_qr(monkeypatch):
+    ''' Check that the expected error is raised in
+    DynBasisFunctions._stub_declarations() if an un-supported quadrature
+    shape is encountered. '''
+    from psyclone.dynamo0p3 import DynBasisFunctions
+    from psyclone.f2pygen import ModuleGen
+    from psyclone import dynamo0p3
+    ast = fpapi.parse(DIFF_BASIS, ignore_comments=False)
+    metadata = DynKernMetadata(ast)
+    kernel = DynKern()
+    kernel.load_meta(metadata)
+    dbasis = DynBasisFunctions(kernel)
+    monkeypatch.setattr(
+        dynamo0p3, "VALID_QUADRATURE_SHAPES",
+        dynamo0p3.VALID_QUADRATURE_SHAPES + ["unsupported-shape"])
+    with pytest.raises(GenerationError) as err:
+        dbasis._stub_declarations(ModuleGen(name="my_mod"))
+    assert ("Quadrature shapes other than GH_QUADRATURE_XYoZ are not yet "
+            "supported - got 'unsupported-shape'" in str(err))
+
+
+def test_dynbasisfns_declns(monkeypatch):
+    ''' Check the various internal errors that
+    DynBasisFunctions._basis_fn_declns can raise. '''
+    from psyclone.dynamo0p3 import DynBasisFunctions
+    ast = fpapi.parse(DIFF_BASIS, ignore_comments=False)
+    metadata = DynKernMetadata(ast)
+    kernel = DynKern()
+    kernel.load_meta(metadata)
+    dbasis = DynBasisFunctions(kernel)
+    dbasis._basis_fns[0]["shape"] = "not-a-shape"
+    with pytest.raises(InternalError) as err:
+        dbasis._basis_fn_declns()
+    assert "Unrecognised evaluator shape: 'not-a-shape'. Should" in str(err)
+    monkeypatch.setattr(dbasis, "_kernel", None)
+    dbasis._basis_fns[0]['type'] = "basis"
+    with pytest.raises(InternalError) as err:
+        dbasis._basis_fn_declns()
+    assert ("basis functions but do not have either a Kernel or an Invoke. "
+            "Should be" in str(err))
+    dbasis._basis_fns[0]['type'] = "diff-basis"
+    with pytest.raises(InternalError) as err:
+        dbasis._basis_fn_declns()
+    assert ("differential basis functions but do not have either a Kernel or "
+            "an Invoke. Should be" in str(err))
+    for fn in dbasis._basis_fns:
```

Comment:
pylint error:

C:1780, 8: Variable name "fn" doesn't conform to snake_case naming style (invalid-name)

Reply:
Fixed.